File: Mono/Tables.iOS/TableAdapter.cs
using System;
using System.Drawing;

using MonoTouch.Foundation;
using MonoTouch.UIKit;

namespace Tables.iOS
{
    public class TableAdapter : NSObject,ITableAdapter
    {
        public ITableAdapterRowSelector RowSelector {get;set;}
        public ITableAdapterRowConfigurator RowConfigurator {get;set;}
        public ITableAdapterRowChanged RowChanged {get;set;}
        private UITableView tv;
        private TableSource td;

		public bool ShouldAdjustTextContentInset=true;

        public TableAdapter(UITableView table=null,Object data=null,ITableAdapterRowConfigurator configs=null) : base()
        {
            td = new TableSource();
            ListView = table;
            RowConfigurator = configs;
            Data = data;
        }

        public UITableView ListView
        {
            get
            {
                return tv;
            }
            set
            {
                if (tv != null)
                {
                    tv.WeakDataSource = null;
                    tv.WeakDelegate = null;
                }
                tv = value;
                if (tv != null)
                {
                    tv.WeakDelegate = this;
                    tv.WeakDataSource = this;
                }
            }
        }

        public Object Data
        {
            get
            {
                return td.Data;             
            }
            set
            {
                td = new TableSource(value);
                ReloadData();
            }
        }

        public void ReloadData()
        {
            if (tv != null)
                tv.ReloadData();
        }

        [Export ("numberOfSectionsInTableView:")]
        public int NumberOfSections (UITableView tableView)
        {
			return td.NumberOfSections();
        }

        [Export ("tableView:numberOfRowsInSection:")]
        public int RowsInSection (UITableView tableView, int section)
        {
            return td.RowsInSection(section);
        }

        static UIFont titleFont = UIFont.SystemFontOfSize(16);
        static UIFont detailFont = UIFont.FromName ("Helvetica", 14);

		public float GetHeightForRow(UITableView tableView, NSIndexPath indexPath)
		{
			var rowType = td.RowType(RowConfigurator,indexPath.Row,indexPath.Section);
			if (rowType == TableRowType.Blurb)
			{
				var name = td.DisplayName(RowConfigurator,indexPath.Row, indexPath.Section);
				var value = td.GetValue(indexPath.Row, indexPath.Section);
				string text = value as string;
				if (text == null) text = "";
				float contentMargin = 10;
				float contentWidth = tableView.Bounds.Width;

				SizeF constraint = new SizeF(contentWidth - (contentMargin * 2), 20000.0f);
				SizeF size = tv.StringSize (text, detailFont, constraint, UILineBreakMode.WordWrap);
				SizeF sizeT = tv.StringSize(name, titleFont);

				float height = Math.Max(size.Height,44.0f) + sizeT.Height;

				return height + (contentMargin * 4);
			}
			return 44;
		}

        [Export("tableView:heightForRowAtIndexPath:")]
        public float HeightForRow(UITableView tableView, NSIndexPath indexPath)
        {
			return GetHeightForRow (tableView, indexPath);
        }

		[Export("tableView:estimatedHeightForRowAtIndexPath:")]
		public float EstimatedHeightForRow(UITableView tableView, NSIndexPath indexPath)
		{
			return GetHeightForRow (tableView, indexPath);
		}

        [Export ("tableView:cellForRowAtIndexPath:")]
        public UITableViewCell GetCell (UITableView tableView, NSIndexPath indexPath)
        {
            var name = td.GetName(indexPath.Row, indexPath.Section);
            var editable = td.Editable(RowConfigurator,indexPath.Row,indexPath.Section);
            var rowType = td.RowType(RowConfigurator,indexPath.Row,indexPath.Section);
            var value = td.GetValue(indexPath.Row, indexPath.Section);

            UITableViewCell cell = tableView.DequeueReusableCell(name);
            if (cell == null )
            {
                //cell = new UITableViewCell(rowType==TableRowType.Blurb?UITableViewCellStyle.Subtitle:UITableViewCellStyle.Value1,name);
				cell = new UITableViewCell(UITableViewCellStyle.Subtitle,name);
                if (rowType==TableRowType.Checkbox)
                {
                    var c = td.RowSetting(RowConfigurator,name);
                    if (c == null || !c.SimpleCheckbox)
                    {
                        var sw = new UISwitch();
                        sw.UserInteractionEnabled = false;
                        cell.AccessoryView = sw;
                    }
                }
				if (rowType == TableRowType.Text)
				{
					var c = td.RowSetting(RowConfigurator,name);
					if (c != null && c.InlineTextEditing)
					{
						var tf = new UITextField(new RectangleF(0,0,160,44));
						tf.UserInteractionEnabled = true;
						tf.BorderStyle = UITextBorderStyle.None;
						tf.AutoresizingMask = UIViewAutoresizing.FlexibleWidth;
						tf.Font = detailFont;
						tf.TextAlignment = UITextAlignment.Right;
						tf.ValueChanged += TextChanged;
						tf.WeakDelegate = this;
						var inp = new TableAdapterInlineTextInputAccessoryView (c, tableView.Frame.Width);
						inp.PreviousButton.TouchUpInside += ClickedPrevious;
						inp.NextButton.TouchUpInside += ClickedNext;
						inp.DismissButton.TouchUpInside += ClickedDismiss;

						tf.InputAccessoryView = inp;
						TableEditor.ConfigureTextControl (c, tf);
						cell.AccessoryView = tf;
					}
				}
                if (rowType == TableRowType.Blurb)
                {
                    cell.TextLabel.Font = titleFont;
                    cell.DetailTextLabel.Lines = 0;
                    cell.DetailTextLabel.Font = detailFont;
					cell.DetailTextLabel.ClipsToBounds = true;
					cell.ClipsToBounds = true;
                }
            }

            cell.TextLabel.Text = td.DisplayName(RowConfigurator,indexPath.Row,indexPath.Section);
            cell.SelectionStyle = editable ? UITableViewCellSelectionStyle.Default : UITableViewCellSelectionStyle.None;

            switch (rowType)
            {
				case TableRowType.Text:
					var vs = value as string;
					var c = td.RowSetting (RowConfigurator, name);
					if (c != null && c.InlineTextEditing)
					{
						var tf = (cell.AccessoryView as UITextField);
						tf.Enabled = editable;
						tf.Text = vs;
						(tf.InputAccessoryView as TableAdapterInlineTextInputAccessoryView).IndexPath = indexPath;
						cell.DetailTextLabel.Text = "";
					}
					else
					{
						cell.DetailTextLabel.Text = vs;
						cell.Accessory = editable ? UITableViewCellAccessory.DisclosureIndicator : UITableViewCellAccessory.None;
					}
                break;
                case TableRowType.Blurb:
                    cell.DetailTextLabel.Text = value as string; 
                    cell.Accessory = editable ? UITableViewCellAccessory.DisclosureIndicator : UITableViewCellAccessory.None;
                break;
                case TableRowType.Checkbox:
                    cell.DetailTextLabel.Text = "";
                    var s = td.RowSetting(RowConfigurator,name);
                    if (s!=null && s.SimpleCheckbox)
                        cell.Accessory = ((bool)value) ? UITableViewCellAccessory.Checkmark : UITableViewCellAccessory.None;
                    else
                        (cell.AccessoryView as UISwitch).On = (bool)value;
                break;
                case TableRowType.DateTime:
                case TableRowType.Date:
                case TableRowType.Time:
                    cell.DetailTextLabel.Text = td.DisplayDate(RowConfigurator, indexPath.Row, indexPath.Section, (DateTime)value, rowType);
                    cell.Accessory = editable ? UITableViewCellAccessory.DisclosureIndicator : UITableViewCellAccessory.None;
                break;
            }

            return cell;
        }

        [Export ("tableView:didSelectRowAtIndexPath:")]
        public void RowSelected (UITableView tableView, NSIndexPath indexPath)
        {
            var name = td.GetName(indexPath.Row,indexPath.Section);
            var value = td.GetValue(indexPath.Row,indexPath.Section);
            if (!td.Editable(RowConfigurator,indexPath.Row,indexPath.Section))
                return;
            var rowType = td.RowType(RowConfigurator,indexPath.Row,indexPath.Section);

            if (RowSelector != null)
            if (RowSelector.DidSelectRow(this,name))
                return;

            TableAdapterRowConfig config = td.RowSetting(RowConfigurator,name);
			if (config != null)
            {
				if (config.Clicked != null)
                {
					config.Clicked(this, null);
                    return;
                }
            }

            switch (rowType)
            {
                case TableRowType.Checkbox:
                    var obj = !(bool)value;
                    td.SetValue(obj, indexPath.Row, indexPath.Section);
                    tableView.ReloadRows(new NSIndexPath[]{ indexPath }, UITableViewRowAnimation.Fade);
                    ChangedValue(name,value,obj);
                break;
                
                case TableRowType.Text:
                case TableRowType.Blurb:
                    var tvc = FirstAvailableViewController;
                    if (tvc != null)
                    {
                        string str = value as string;

						if (config != null && config.InlineTextEditing && rowType==TableRowType.Text)
						{
							var cell = tableView.CellAt (indexPath);
							var tf = (cell.AccessoryView as UITextField);
							tf.BecomeFirstResponder ();
						}
						else
						{
							var dname = td.DisplayName (RowConfigurator, indexPath.Row, indexPath.Section);
							var textEditor = new TableTextEditor (rowType, dname, str, delegate(string changedString)
							{
								td.SetValue (changedString, indexPath.Row, indexPath.Section);
								ReloadData ();
							});
							textEditor.ShouldAdjustTextContentInset = ShouldAdjustTextContentInset;
							textEditor.Configure (config);

							if (tvc.NavigationController == null)
								tvc.PresentViewController (new UINavigationController (textEditor), true, null);
							else
								tvc.NavigationController.PushViewController (textEditor, true);
						}
                    }
                break;
                case TableRowType.Date:
                case TableRowType.Time:
                case TableRowType.DateTime:
                    var dvc = FirstAvailableViewController;
                    if (dvc != null)
                    {
                        DateTime str = (DateTime)value;
                        var dname = td.DisplayName(RowConfigurator, indexPath.Row, indexPath.Section);
                        UIDatePickerMode mode = UIDatePickerMode.DateAndTime;
                        if (rowType == TableRowType.Date)
                            mode = UIDatePickerMode.Date;
                        else if (rowType == TableRowType.Time)
                            mode = UIDatePickerMode.Time;

						var dateEditor = new TableTimeEditor (dname, str, mode, delegate(DateTime changedDate)
						{
							td.SetValue (changedDate, indexPath.Row, indexPath.Section);
							ReloadData ();
						});
						if (dvc.NavigationController == null)
							dvc.PresentViewController(new UINavigationController(dateEditor), true, null);
						else
							dvc.NavigationController.PushViewController (dateEditor, true);							
                    }
                break;
            }
        }

        private void ChangedValue(string rowName,object oldValue,object newValue)
        {
            if (RowChanged != null)
                RowChanged.RowChanged(this, rowName, oldValue, newValue);
        }

        private UIViewController FirstAvailableViewController
        {
            get
            {
                return TableAdapter.TraverseResponderChainForViewController(tv) as UIViewController;
            }
        }

        static private UIResponder TraverseResponderChainForViewController(UIView v)
        {
            UIResponder responder = v.NextResponder;
            if (responder is UIViewController)
            {
                return responder;
            }
            else if (responder is UIView)
            {
                return TraverseResponderChainForViewController(responder as UIView);
            }
            else
            {
                return null;
            }
        }

		#region Inline Text Editing

		NSIndexPath NextIndexPath(NSIndexPath indexPath)
		{
			int numOfSections = NumberOfSections(tv);				
			int nextSection = ((indexPath.Section + 1) % numOfSections);

			if ((indexPath.Row + 1) == RowsInSection(tv,indexPath.Section)) 
			{
				return NSIndexPath.FromRowSection(0,nextSection);
			} 
			else
			{
				return NSIndexPath.FromRowSection((indexPath.Row + 1),indexPath.Section);
			}
		}

		NSIndexPath PreviousIndexPath(NSIndexPath indexPath)
		{
			int numOfSections = NumberOfSections(tv);				
			int nextSection = ((indexPath.Section - 1) % numOfSections);

			if ((indexPath.Row - 1) < 0) 
			{
				return NSIndexPath.FromRowSection(0,nextSection);
			} 
			else
			{
				return NSIndexPath.FromRowSection((indexPath.Row - 1),indexPath.Section);
			}
		}

		void ClickedPrevious (object sender, EventArgs e)
		{
			NSIndexPath indexPath = ((sender as UIView).Superview as TableAdapterInlineTextInputAccessoryView).IndexPath;
			NSIndexPath nextIndexPath = PreviousIndexPath (indexPath);
			tv.SelectRow (nextIndexPath, true, UITableViewScrollPosition.Top);
			RowSelected (tv, nextIndexPath);
		}

		void ClickedNext (object sender, EventArgs e)
		{
			NSIndexPath indexPath = ((sender as UIView).Superview as TableAdapterInlineTextInputAccessoryView).IndexPath;
			NSIndexPath nextIndexPath = NextIndexPath (indexPath);
			tv.SelectRow (nextIndexPath, true, UITableViewScrollPosition.Top);
			RowSelected (tv, nextIndexPath);
		}

		void ClickedDismiss (object sender, EventArgs e)
		{
			UIApplication.SharedApplication.KeyWindow.EndEditing(true);
		}

		void TextChanged (object sender, EventArgs e)
		{
			UITextField tf = sender as UITextField;
			NSIndexPath indexPath = (tf.InputAccessoryView as TableAdapterInlineTextInputAccessoryView).IndexPath;
			td.SetValue (tf.Text, indexPath.Row, indexPath.Section);
		}

		[Export ("textFieldShouldReturn:")]
		public bool TextFieldShouldReturn (UITextField tf)
		{
			NSIndexPath indexPath = (tf.InputAccessoryView as TableAdapterInlineTextInputAccessoryView).IndexPath;
			NSIndexPath nextIndexPath = NextIndexPath (indexPath);
			tv.SelectRow (nextIndexPath, true, UITableViewScrollPosition.Top);
			RowSelected (tv, nextIndexPath);
			return true;
		}

		#endregion
    }
}

File: objc/Tables/TableAdapter.m
//
//  TableAdapter.m
//  tables
//
//  Created by Aleksander Slater on 07/09/2014.
//  Copyright (c) 2014 Davincium. All rights reserved.
//

#import "TableAdapter.h"
#import "TableSource.h"

#import "TableAdapterCell.h"
#import "TableEditor.h"
#import "TableTextEditor.h"
#import "TableSingleChoiceEditor.h"
#import "TableTimeEditor.h"

#define TITLE_FONT [UIFont systemFontOfSize:16]
#define DETAIL_FONT [UIFont fontWithName:@"Helvetica" size:14]

@interface TableAdapter()
@property(nonatomic,strong) UITableView *tv;
@property(nonatomic,strong) TableSource *td;
@end

@implementation TableAdapter

- (id)initWithTable:(UITableView*)table andData:(NSObject *)data usingConfigs:(id<ITableAdapterRowConfigurator>)configs
{
    self = [super init];
    if (self)
    {
        self.shouldAdjustTextContentInset=YES;
        self.detailTextColor = [UIColor darkGrayColor];
        
        self.td = [[TableSource alloc] initWithData:data];
        self.rowConfigurator = configs;
        self.data = data;
        [self setTableView:table];
    }
    return self;
}

- (id)initWithTable:(UITableView*)table andData:(NSObject *)data
{
    self = [self initWithTable:table andData:data usingConfigs:nil];
    if (self)
    {
        
    }
    return self;
}

- (void)setTableView:(UITableView*)table
{
    self.tv.delegate = nil;
    self.tv.dataSource = nil;
    self.tv = nil;
    self.tv = table;
    self.tv.delegate = self;
    self.tv.dataSource = self;
}

- (void)setData:(NSObject *)data
{
    self.td = [[TableSource alloc] initWithData:data];
    [self reloadData];
}

- (NSObject*)data
{
    return self.td.data;
}

- (void)reloadData
{
    [self.tv reloadData];
}

#pragma mark - UITableView

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.td.numberOfSections;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.td rowsInSection:section];
}

- (CGFloat)getHeightForRow:(UITableView*)tableView atIndexPath:(NSIndexPath*)indexPath
{
    TableRowType rowType = [self.td rowType:self.rowConfigurator row:indexPath.row section:indexPath.section];
    if (rowType == TableRowTypeBlurb)
    {
        NSString* name = [self.td displayName:self.rowConfigurator row:indexPath.row section:indexPath.section];
        NSString* text = [self.td getValueStringAtRow:indexPath.row andSection:indexPath.section];
        CGFloat contentMargin = 10;
        CGFloat contentWidth = tableView.bounds.size.width;
        CGSize constraint = CGSizeMake(contentWidth - (contentMargin * 2), 20000.0f);
        
        CGSize size = [TextHelper stringSize:text font:DETAIL_FONT constraint:constraint lineBreakMode:NSLineBreakByWordWrapping];
        CGSize sizeT = [TextHelper stringSize:name font:TITLE_FONT];
        CGFloat height = MAX(size.height,44.0f) + sizeT.height;
        return height + (contentMargin * 4);
    }
    return 44;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return [self getHeightForRow:tableView atIndexPath:indexPath];
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return [self getHeightForRow:tableView atIndexPath:indexPath];
}

- (UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *name = [self.td rowName:indexPath.row andSection:indexPath.section];
    BOOL editable = [self.td rowEditable:self.rowConfigurator row:indexPath.row section:indexPath.section];
    TableRowType rowType = [self.td rowType:self.rowConfigurator row:indexPath.row section:indexPath.section];
    NSObject *value = [self.td getValueAtRow:indexPath.row andSection:indexPath.section];
    TableAdapterCell *cell = (TableAdapterCell*)[tableView dequeueReusableCellWithIdentifier:name];
    if (cell == nil )
    {
        cell = [[TableAdapterCell alloc] initWithStyle:(rowType==TableRowTypeBlurb?UITableViewCellStyleSubtitle:UITableViewCellStyleValue1) reuseIdentifier:name];
        cell.accessoryType = UITableViewCellAccessoryNone;
        //cell = new TableAdapterCell(UITableViewCellStyle.Subtitle,name);
        switch (rowType)
        {
            case TableRowTypeCheckbox:
            {
                TableAdapterRowConfig*c=[self.td rowSetting:self.rowConfigurator propertyName:name];
                if (c == nil || !c.simpleCheckbox)
                {
                    UISwitch* sw = [[UISwitch alloc] init];
                    sw.userInteractionEnabled = false;
                    cell.accessoryView = sw;
                }
            } break;
            case TableRowTypeText:
            {
                TableAdapterRowConfig*c=[self.td rowSetting:self.rowConfigurator propertyName:name];
                if (c != nil && c.inlineTextEditing)
                {
                    UITextField* tf = [[UITextField alloc]initWithFrame:CGRectMake(0, 0, 160, 44)];
                    tf.userInteractionEnabled = true;
                    tf.borderStyle = UITextBorderStyleNone;
                    tf.autoresizingMask = UIViewAutoresizingFlexibleWidth;
                    tf.font = DETAIL_FONT;
                    tf.textColor = self.detailTextColor;
                    tf.textAlignment = NSTextAlignmentRight;
                    tf.delegate = self;
                    [tf addTarget:self action:@selector(textFieldChanged:) forControlEvents:UIControlEventEditingChanged];
                    TableAdapterInlineTextInputAccessoryView* inp = [[TableAdapterInlineTextInputAccessoryView alloc] initWithConfig:c andWidth:tableView.frame.size.width];
                    [inp.previousButton addTarget:self action:@selector(clickedPrevious:) forControlEvents:UIControlEventTouchUpInside];
                    [inp.nextButton addTarget:self action:@selector(clickedNext:) forControlEvents:UIControlEventTouchUpInside];
                    [inp.dismissButton addTarget:self action:@selector(clickedDismiss:) forControlEvents:UIControlEventTouchUpInside];
                    tf.inputAccessoryView = inp;
                    [TableEditor configureTextControl:c control:tf];
                    cell.accessoryView = tf;
                }
                else
                {
                    cell.detailTextLabel.textColor = self.detailTextColor;
                }
            } break;
            case TableRowTypeBlurb:
            {
                cell.textLabel.font = TITLE_FONT;
                cell.detailTextLabel.numberOfLines = 0;
                cell.detailTextLabel.font = DETAIL_FONT;
                cell.detailTextLabel.clipsToBounds = true;
                cell.detailTextLabel.textColor = self.detailTextColor;
                cell.clipsToBounds = true;
            } break;
            default:
            {
                cell.detailTextLabel.textColor = self.detailTextColor;
            } break;
        }
    }

    cell.textLabel.text = [self.td displayName:self.rowConfigurator row:indexPath.row section:indexPath.section];
    cell.selectionStyle = editable ? UITableViewCellSelectionStyleDefault : UITableViewCellSelectionStyleNone;
    switch (rowType)
    {
        case TableRowTypeText:
        {
            NSString* vs = [self.td getValueStringAtRow:indexPath.row andSection:indexPath.section];
            TableAdapterRowConfig*c=[self.td rowSetting:self.rowConfigurator propertyName:name];
            if (c != nil && c.inlineTextEditing)
            {
                UITextField *tf = (UITextField*)cell.accessoryView;
                tf.enabled = editable;
                tf.text = vs;
                tf.secureTextEntry = c.secureTextEditing;
                TableAdapterInlineTextInputAccessoryView*inp = (TableAdapterInlineTextInputAccessoryView*)tf.inputAccessoryView;
                inp.indexPath = indexPath;
                cell.detailTextLabel.text = @"";
            }
            else
            {
                cell.detailTextLabel.text = c != nil && c.secureTextEditing ? [TextHelper scrambledText:vs] : vs;
                cell.accessoryType = editable ? UITableViewCellAccessoryDisclosureIndicator : UITableViewCellAccessoryNone;
            }
        } break;

        case TableRowTypeBlurb:
        {
            NSString* vs = [self.td getValueStringAtRow:indexPath.row andSection:indexPath.section];
            TableAdapterRowConfig*c=[self.td rowSetting:self.rowConfigurator propertyName:name];
            cell.detailTextLabel.text = c != nil && c.secureTextEditing ? [TextHelper scrambledText:vs] : vs;
            cell.accessoryType = editable ? UITableViewCellAccessoryDisclosureIndicator : UITableViewCellAccessoryNone;
        } break;
            
        case TableRowTypeSingleChoiceList:
        {
            NSString* choiceString = nil;
            if (value != nil)
                choiceString = [value description];
            cell.detailTextLabel.text = choiceString;
            cell.accessoryType = editable ? UITableViewCellAccessoryDisclosureIndicator : UITableViewCellAccessoryNone;
        } break;
        
        case TableRowTypeCheckbox:
        {
            cell.detailTextLabel.text = @"";
            BOOL isOn = [self.td getValueBoolAtRow:indexPath.row andSection:indexPath.section];
            TableAdapterRowConfig*s=[self.td rowSetting:self.rowConfigurator propertyName:name];
            if (s != nil && s.simpleCheckbox)
                cell.accessoryType = isOn ? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
            else
            {
                UISwitch *sw = (UISwitch*)cell.accessoryView;
                sw.on = isOn;
            }
        } break;

        case TableRowTypeDateTime:
        case TableRowTypeDate:
        case TableRowTypeTime:
        {
            NSDate *date = [self.td getValueDateAtRow:indexPath.row andSection:indexPath.section];
            [self.td displayDate:self.rowConfigurator row:indexPath.row section:indexPath.section date:date rowType:rowType];
            cell.accessoryType = editable ? UITableViewCellAccessoryDisclosureIndicator : UITableViewCellAccessoryNone;
        } break;
            
        case TableRowTypeUnknown:
        {
            
        } break;
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *name = [self.td rowName:indexPath.row andSection:indexPath.section];
    BOOL editable = [self.td rowEditable:self.rowConfigurator row:indexPath.row section:indexPath.section];
    if (!editable)
        return;
    TableRowType rowType = [self.td rowType:self.rowConfigurator row:indexPath.row section:indexPath.section];
    if (self.rowSelector != nil)
        if ([self.rowSelector  tableAdapter:self didSelectRow:name])
                return;
    TableAdapterRowConfig*config=[self.td rowSetting:self.rowConfigurator propertyName:name];
    if (config != nil)
    {
        if (config.clicked!=nil)
        {
            [self performSelectorOnMainThread:config.clicked withObject:nil waitUntilDone:YES];
            return;
        }
    }
    switch (rowType)
    {
        case TableRowTypeCheckbox:
        {
            NSNumber* number = [self.td getValueNumberAtRow:indexPath.row andSection:indexPath.section];
            BOOL isOn = ![number boolValue];
            NSNumber *newValue = @(isOn);
            [self.td setValue:newValue atRow:indexPath.row andSection:indexPath.section];
            [tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
            [self changedValue:name oldValue:number newValue:newValue];
        } break;
        case TableRowTypeText:
        case TableRowTypeBlurb:
        {
            UIViewController* tvc = self.firstAvailableViewController;
            if (tvc != nil)
            {
                NSString* str = [self.td getValueStringAtRow:indexPath.row andSection:indexPath.section];
                if (config != nil && config.inlineTextEditing && rowType==TableRowTypeText)
                {
                    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
                    UITextField* tf = (UITextField*)cell.accessoryView;
                    [tf becomeFirstResponder];
                }
                else
                {
                    NSString* dname = [self.td displayName:self.rowConfigurator row:indexPath.row section:indexPath.section];
                    TableTextEditor* textEditor = [[TableTextEditor alloc] initWithRowType:rowType title:dname value:str changed:^(NSString*changedString)
                    {
                        [self.td setValue:changedString atRow:indexPath.row andSection:indexPath.section];
                        [self reloadData];
                    }];
                    [textEditor setShouldAdjustTextContentInset:self.shouldAdjustTextContentInset];
                    [textEditor configure:config];
                    [TableAdapter presentScreen:textEditor fromParent:tvc];
                }
            }
        } break;
        case TableRowTypeSingleChoiceList:
        {
            UIViewController *sclvc = self.firstAvailableViewController;
            if (sclvc != nil)
            {
                NSString* dname = [self.td displayName:self.rowConfigurator row:indexPath.row section:indexPath.section];
                NSObject* value = [self.td getValueAtRow:indexPath.row andSection:indexPath.section];
                TableSingleChoiceEditor* singleChoiceEditor = [[TableSingleChoiceEditor alloc] initWithRowType:rowType
                                                                                                         title:dname
                                                                                                  chosenOption:value
                                                                                                        config:config
                                                                                                       changed:^(NSObject* changedChoice)
                {
                    [self.td setValue:changedChoice atRow:indexPath.row andSection:indexPath.section];
                    [self reloadData];
                }];
                [TableAdapter presentScreen:singleChoiceEditor fromParent:sclvc];
            }
        }  break;
        case TableRowTypeDate:
        case TableRowTypeTime:
        case TableRowTypeDateTime:
        {
            UIViewController *dvc = self.firstAvailableViewController;
            if (dvc != nil)
            {
                NSString* dname = [self.td displayName:self.rowConfigurator row:indexPath.row section:indexPath.section];
                NSDate* value = [self.td getValueDateAtRow:indexPath.row andSection:indexPath.section];
                UIDatePickerMode mode = UIDatePickerModeDateAndTime;
                if (rowType == TableRowTypeDate)
                    mode = UIDatePickerModeDate;
                else if (rowType == TableRowTypeTime)
                    mode = UIDatePickerModeTime;
                TableTimeEditor* dateEditor = [[TableTimeEditor alloc] initWithTitle:dname value:value mode:mode changed:^(NSDate*changedDate)
                {
                    [self.td setValue:changedDate atRow:indexPath.row andSection:indexPath.section];
                    [self reloadData];
                }];
                [TableAdapter presentScreen:dateEditor fromParent:dvc];
            }
        } break;
            
        case TableRowTypeUnknown:
        {
            
        } break;
    }
}

- (void)changedValue:(NSString*)rowName oldValue:(NSObject*)oldValue newValue:(NSObject*)newValue
{
    [self.rowChanged rowChanged:self rowName:rowName oldValue:oldValue newValue:newValue];
}

+ (void)presentScreen:(UIViewController*)screen fromParent:(UIViewController*)parent
{
    if (parent.navigationController == nil)
        [parent presentViewController:[[UINavigationController alloc]initWithRootViewController:screen] animated:YES completion:nil];
    else
        [parent.navigationController pushViewController:screen animated:YES];
}

- (UIViewController*)firstAvailableViewController
{
    return (UIViewController*)[TableAdapter traverseResponderChainForViewController:self.tv];
}

+ (UIResponder*)traverseResponderChainForViewController:(UIView*)v
{
    UIResponder *responder = [v nextResponder];
    if ([responder isKindOfClass:[UIViewController class]])
    {
        return responder;
    }
    else if ([responder isKindOfClass:[UIView class]])
    {
        return [TableAdapter traverseResponderChainForViewController:(UIView*)responder];
    }
    else
    {
        return nil;
    }
}

#pragma mark - Inline Text Editing

- (NSIndexPath*)nextIndexPath:(NSIndexPath*)indexPath
{
    NSInteger numOfSections = [self numberOfSectionsInTableView:self.tv];
    NSInteger nextSection = ((indexPath.section + 1) % numOfSections);
    
    if ((indexPath.row + 1) == [self tableView:self.tv numberOfRowsInSection:indexPath.section])
    {
        return [NSIndexPath indexPathForRow:0 inSection:nextSection];
    }
    else
    {
        return [NSIndexPath indexPathForRow:indexPath.row + 1 inSection:indexPath.section];
    }
}

- (NSIndexPath*)previousIndexPath:(NSIndexPath*)indexPath
{
    NSInteger numOfSections = [self numberOfSectionsInTableView:self.tv];
    NSInteger nextSection = ((indexPath.section - 1) % numOfSections);
    if ((indexPath.row - 1) < 0)
    {
        return [NSIndexPath indexPathForRow:0 inSection:nextSection];
    }
    else
    {
        return [NSIndexPath indexPathForRow:indexPath.row - 1 inSection:indexPath.section];
    }
}

- (IBAction)clickedPrevious:(id)sender
{
    UIView *v = (UIView *)sender;
    TableAdapterInlineTextInputAccessoryView *ac = (TableAdapterInlineTextInputAccessoryView*)[v superview];
    NSIndexPath *indexPath =ac.indexPath;
    NSIndexPath *nextIndexPath = [self previousIndexPath:indexPath];
    [self.tv selectRowAtIndexPath:nextIndexPath animated:YES scrollPosition:UITableViewScrollPositionTop];
    [self tableView:self.tv didSelectRowAtIndexPath:nextIndexPath];
}

- (IBAction)clickedNext:(id)sender
{
    UIView *v = (UIView *)sender;
    TableAdapterInlineTextInputAccessoryView *ac = (TableAdapterInlineTextInputAccessoryView*)[v superview];
    NSIndexPath *indexPath =ac.indexPath;
    NSIndexPath *nextIndexPath = [self nextIndexPath:indexPath];
    [self.tv selectRowAtIndexPath:nextIndexPath animated:YES scrollPosition:UITableViewScrollPositionTop];
    [self tableView:self.tv didSelectRowAtIndexPath:nextIndexPath];
}

- (IBAction)clickedDismiss:(id)sender
{
    [[[UIApplication sharedApplication] keyWindow] endEditing:YES];
}

- (void)textFieldChanged:(UITextField*)tf
{
    TableAdapterInlineTextInputAccessoryView*inp = (TableAdapterInlineTextInputAccessoryView*)tf.inputAccessoryView;
    NSIndexPath *indexPath = inp.indexPath;
    [self.td setValue:tf.text atRow:indexPath.row andSection:indexPath.section];
}

-(BOOL)textFieldShouldReturn:(UITextField *)tf
{
    TableAdapterInlineTextInputAccessoryView*inp = (TableAdapterInlineTextInputAccessoryView*)tf.inputAccessoryView;
    NSIndexPath *indexPath = inp.indexPath;
    NSIndexPath *nextIndexPath = [self nextIndexPath:indexPath];
    [self.tv selectRowAtIndexPath:nextIndexPath animated:YES scrollPosition:UITableViewScrollPositionTop];
    [self tableView:self.tv didSelectRowAtIndexPath:nextIndexPath];
    return YES;
}

//- (BOOL)textField:(UITextField *)tf shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)replacement
//{
//    NSString* text = tf.text;
//    tf.Text = text.Substring (0, range.Location) + replacement + text.Substring (range.Location + range.Length);
//    TextFieldChanged (tf);
//    return false;
//}

@end
